students/Nick_Haury/Lesson06/mailroom4.py

```diff
@@ -0,0 +1,317 @@
+                         'donation for:\n>>')
+
+        if ty_input.lower() == "list":
+            print("\nCurrent Donors: ")
+            for row in get_donor_list():
+                print(row)
+            print()
+        else:
+            try:
+                donation_amount = float(input("\nPlease enter donation amount:"
+                                              "\n>>"))
+                if donation_amount <= 0:
+                    raise ValueError
+            except ValueError:
+                print("\nPlease only enter a single positive number for "
+                      "donation amount\n")
+            else:
+                add_donation(ty_input, donation_amount)
+                print(
+                    "\n", create_email_text(ty_input, donation_amount), sep=""
+                    )
+                break
+
+
+def get_donor_list(donors=donors):
+    return list(donors.keys())
+
+
+def create_email_text(donor_name, donation_amount):
+    '''
+    Takes a donor's name and donation amount as parameters, and then creates
+    an email template thanking them for their donation
+    '''
+
+    return(f"Dear {donor_name},\n\n"
+           f"It is with incredible gratitude that we accept your wonderfully "
+           f"generous donation of ${donation_amount:,.2f}.  Your "
+           "contribution will truly make a difference in the path forward "
+           "towards funding our common goal."
+           "\n\nEver Greatefully Yours,\n\n"
+           "X" + ("_" * 20) + "\n")
+
+
+def add_donation(donor_name, donation_amount, donors=donors):
+    '''
+    Adds donation amount to donor's list of donations.  If donor does not exist
+    yet, they are created and added to the donor dictionary.
+    '''
+    donors.setdefault(donor_name, []).append(donation_amount)
+
+
+def create_report():
+    '''
+    Prints a list of donors, sorted by total historical donation
+    amount.  Includes donor name, total donated, number of donations,
+    and average donation amount.
+    '''
+    print("\nDonor Name" + " "*15 + "|  Total Given  | Num Gifts |   "
+          "Average Gift")
+    print("- "*36)
+    for row in get_report_text():
+        print(row)
+    print()
+
+
+def get_report_text(donors=donors):
+
+    def sort_key(donor):
+        # sorting by the sum of donations
+        return sum(donors[donor])
+
+    keys_by_sum = sorted(donors, key=sort_key, reverse=True)
+    text_list = []
+    for key in keys_by_sum:
+        donor_name = key
+        donor_sum = sum(donors[key])
+        donor_count = len(donors[key])
+        donor_average = donor_sum/donor_count
+        text_list.append(f"{donor_name:26}${donor_sum:14,.2f}{donor_count:11}"
+                         f"  ${donor_average:16,.2f}")
+    return text_list
+
+
+def write_letters(donors=donors):
+    for key in donors:
+        with open(f'{key}.txt', 'w') as f:
+            f.write(create_email_text(key, donors[key][-1]))
+    print()
+
+# Run unit tests
```

Comment:
Wasn't sure if the tests should be broken out into a separate file or not.  Prompt didn't seem to indicate either way, so kept them within the main file.

Reply:
yes you do want to keep your tests in a separate file